### PR TITLE
[feature] #2103: support querying for blocks and transactions

### PR DESCRIPTION
**Pull request by QuentinI**
### Description of the Change

Added `FindAllBlocks` and `FindAllTransactions` queries as described in #2103 

### Issue

No way to query for e.g. block explorer to query for older blocks

### Benefits

Support querying for list of all blocks and transactions

### Possible Drawbacks

As implemented in this PR, FindAllBlocks query introduces a block representation in `iroha_data_schema` parallel to one already present in `iroha_core`.

### CODE REVIEW

File: data_model/src/events/mod.rs
Comment (QuentinI) [inline data_model/src/events/mod.rs:91]: `Serialize` and `Deserialize` was missing from events. I added it, but I wonder if it was on purpose?

**Review by appetrosyan — APPROVED** [review]

File: data_model/src/block_value.rs
Comment (QuentinI) [inline data_model/src/block_value.rs:32]: Since I can't use `VersionedCommitedBlock` here and `HashOf<BlockValue>` would be different to `HashOf<VersionedCommitedBlock>` it's just naked `Hash`. I don't think there's any good solution to be found without moving definition of blocks to separate crate entirely.

File: data_model/src/lib.rs
Comment (mversic) [inline data_model/src/lib.rs:433]: hm, where is +1 coming from?

File: data_model/src/block_value.rs
Comment (QuentinI) [inline data_model/src/block_value.rs:60]: I'm not sure I fully understand the semantics here, so I'm just adding lengths of everything having a length here :)

File: data_model/src/block_value.rs
Comment (mversic) [reply data_model/src/block_value.rs:32]: yes, actually blocks should be moved to the `data_model` crate from the `core` crate. Because as far as I can see you are duplicating the block definition

File: data_model/src/lib.rs
Comment (QuentinI) [reply data_model/src/lib.rs:433]: From counting the block itself, I guess?
As I said, I'm not sure I fully understand the semantics of this.

File: data_model/src/block_value.rs
Comment (mversic) [reply data_model/src/block_value.rs:32]: while this is what I would like to see the most I'm not sure everyone would agree

File: core/src/smartcontracts/isi/block.rs
Comment (mversic) [inline core/src/smartcontracts/isi/block.rs:10]: please don't use `#[log]`. It causes deadlocks as far as I'm aware of

File: core/src/smartcontracts/isi/tx.rs
Comment (mversic) [inline core/src/smartcontracts/isi/tx.rs:11]: please don't use `#[log]`. It causes deadlocks as far as I'm aware of

File: data_model/src/block_value.rs
Comment (appetrosyan) [reply data_model/src/block_value.rs:32]: I think that Block -> data_model is beyond the scope of a last minute change.

**Review by mversic — CHANGES_REQUESTED** [review]

**Comment by QuentinI** [comment]
@mversic re: https://github.com/hyperledger/iroha/pull/2210#discussion_r872538610
I agree that duplicating the block definition here is not ideal.
It's a quick solution we agreed on with @appetrosyan , as far as I understand it's planned to move block definitions to their own crate or do some other crate structure refactoring to that effect.

File: core/src/smartcontracts/isi/tx.rs
Comment (appetrosyan) [reply core/src/smartcontracts/isi/tx.rs:11]: Doesnt matter, I'll remove it in #2186 anyway.

File: data_model/src/block_value.rs
Comment (QuentinI) [reply data_model/src/block_value.rs:53]: I removed fields based on recommendations of @appetrosyan, although I may have mixed something up.
Not sure why I merged transactions into one vector though, I'll try to recall if I had any sensible reasoning for it and revert if not.

File: data_model/src/events/mod.rs
Comment (appetrosyan) [reply data_model/src/events/mod.rs:91]: We might need the two soon

**Review by appetrosyan — APPROVED** [review]

File: data_model/src/transaction.rs
Comment (mversic) [inline data_model/src/transaction.rs:427]: do we need it encoded as both scale and json now?

Comment (mversic) [reply data_model/src/transaction.rs:427]: and how come `PartialEq` and `Eq` are needed?

File: core/src/smartcontracts/isi/tx.rs
Comment (mversic) [reply core/src/smartcontracts/isi/tx.rs:10]: I'm think @appetrosyan PR got merged first, so this would reintroduce the `#[log]`

File: data_model/src/query.rs
Comment (mversic) [inline data_model/src/query.rs:1273]: I think we should support n last or from hash filtering in the `FindAllTransactions` query

File: core/src/wsv.rs
Comment (mversic) [inline core/src/wsv.rs:688]: I think transactions should be grouped by block number, i.e. function prototype should be something like this:
```suggestion
    pub fn transaction_values(&self) -> Vec<(u32, Vec<TransactionValue>)> {
```

or block number can be implicit in the position in the vector:
```suggestion
    pub fn transaction_values(&self) -> Vec<Vec<TransactionValue>> {
```

I somehow dislike the idea of mixing valid and rejected transactions so maybe this format could be favored:
 ```suggestion
    pub fn transaction_values(&self) -> Vec<(Vec<ValidTransaction>, Vec<RejectedTransaction>)> {
```

File: data_model/src/query.rs
Comment (QuentinI) [reply data_model/src/query.rs:1273]: I don't know much about the use-cases, so I'll defer to your judgement here.

File: core/src/wsv.rs
Comment (QuentinI) [reply core/src/wsv.rs:688]: Oh, this function isn't actually needed anymore since I split the transactions into valid and rejected, so I should actually remove it.

That said, having block numbers in `FindAllTransactions` output makes sense to me, but again, not sure what the actual use-cases of the query are.

File: data_model/src/query.rs
Comment (appetrosyan) [reply data_model/src/query.rs:1273]: `from_hash` — good idea. `n_last` could be emulated with pagination, so I'd add a new ticket to the backlog.

Comment (mversic) [reply data_model/src/query.rs:1273]: yes, should be in a separate ticket. I'll resolve this then

File: core/src/wsv.rs
Comment (mversic) [reply core/src/wsv.rs:688]: that's true, I too don't know the use-case of this feature so you could just open a ticket for this

**Review by mversic — DISMISSED** [review]

**Comment by QuentinI** [comment]
@appetrosyan @mversic  I applied minor suggestions and rebased.

Do I understand correctly that after this being merged I should open new issues for `from_hash` and block numbering in `FindAllTransactions` ?

**Comment by mversic** [comment]
Yeah, let's do that

**Review by appetrosyan — APPROVED** [review]

**Review by mversic — APPROVED** [review]

**Pull request merged**
